### PR TITLE
[WIP] Build Systems a la Carte

### PR DESCRIPTION
**Pull request by pakoito**
Adds an (incomplete, modified version) of https://www.microsoft.com/en-us/research/uploads/prod/2018/03/build-systems.pdf

### CODE REVIEW

**Comment by pakoito** [comment]
Those commits are showing in all of my pull requests. I'm starting to suspect that my repo is fucked.

**Comment by pakoito** [comment]
Please do not merge it yet. I'm making changes.

File: modules/docs/arrow-examples/src/test/kotlin/arrow/BuildSystemsALaCarte.kt [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+          val newValue: V = !newTask { just(store.getValue(it)) }
+          store.putValue(this@topological, currTarget, newValue)
+        })
+      }
+    }
+  }
+}
+
+// I didn't like the original version using State, so I FUCKED IT UP *sigh*
+fun <F, I, K, V> suspending(): Scheduler<F, I, I, K, V> = { rebuilder: Rebuilder<F, I, K, V> ->
```
Comment (raulraja) [inline modules/docs/arrow-examples/src/test/kotlin/arrow/BuildSystemsALaCarte.kt]: can probably use `Ref` instead of state and pass those around

**Comment by devesh-shetty** [comment]
@pakoito is it done or do you want me to finish this up?

**Comment by pakoito** [comment]
I got stuck because we lack Existential types and this is as far as I was able to drive it (and some parts are even broken). My plan was to keep it this way, if you'd like to clean it up with some tests and merge it I'd say go for it :D

**Comment by snowleopard** [comment]
@pakoito @devesh-shetty I'm working an extended version of the paper right now, and have just linked to this Kotlin implementation as:

Estevez, Paco, et al. (2019). Translation of Build Systems à la Carte to Kotlin.
https://web.archive.org/web/20191020002058/https://github.com/arrow-kt/arrow/blob/paco-tsalc/modules/docs/arrow-examples/src/test/kotlin/arrow/BuildSystemsALaCarte.kt.

I'd like to replace the `et al` bit with an accurate list of contributors, could you please list all authors?

By the way, I hope you don't mind the citation. Here is the context:

https://github.com/snowleopard/build/blob/de30db8a55e17e3cbd9d78857d4f3d634134a538/papers/jfp/7-experience.tex#L40-L47

**Comment by pakoito** [comment]
Oh yes, I'd be honored to appear on the extended paper! My first citation too! 

`Devesh Shetty` (@devesh-shetty in this Issue) is the other person who contributed to it!

**Comment by snowleopard** [comment]
@pakoito Great, I've added @devesh-shetty. Thanks for your translation!

**Comment by devesh-shetty** [comment]
> @pakoito Great, I've added @devesh-shetty. Thanks for your translation!

Thanks a ton! Thats amazing!

**Comment by devesh-shetty** [comment]
> I got stuck because we lack Existential types and this is as far as I was able to drive it (and some parts are even broken). My plan was to keep it this way, if you'd like to clean it up with some tests and merge it I'd say go for it :D

Aye captain!

**Comment by rachelcarmena** [comment]
Hi! Libraries have been moved to the new repositories. Please, find the steps here: https://github.com/arrow-kt/arrow/blob/master/docs/move-to-multi-repo/how-to-move-a-pr.md and let me know if I can help you to move this PR to the correspondent repository. Thanks for your collaboration!!

**Pull request closed without merging**